test/framework/src/main/java/org/elasticsearch/transport/nio/MockNioTransport.java

```diff
@@ -311,4 +323,80 @@ public void sendMessage(BytesReference reference, ActionListener<Void> listener)
         }
     }
+
+    static final class TransportThreadWatchdog {
+
+        private static final long WARN_THRESHOLD = TimeUnit.MILLISECONDS.toNanos(150);
+
+        // Only check every 2s to not flood the logs on a blocked thread.
+        // We mostly care about long blocks and not random slowness anyway and in tests would randomly catch slow operations that block for
+        // less than 2s eventually.
+        private static final TimeValue CHECK_INTERVAL = TimeValue.timeValueSeconds(2);
+
+        private final ThreadPool threadPool;
+        private final AtomicInteger registered = new AtomicInteger(0);
+        private final AtomicBoolean running = new AtomicBoolean(false);
+        private final ConcurrentHashMap<Thread, Long> registry = new ConcurrentHashMap<>();
+
+        private volatile Scheduler.ScheduledCancellable cancellable;
+
+        TransportThreadWatchdog(ThreadPool threadPool) {
+            this.threadPool = threadPool;
+        }
+
+        public boolean register() {
+            registered.getAndIncrement();
+            Long previousValue = registry.put(Thread.currentThread(), threadPool.relativeTimeInNanos());
+            if (previousValue != null) {
+                return false;
+            }
+            if (running.compareAndSet(false, true)) {
+                assert cancellable == null;
+                cancellable = threadPool.schedule(this::logLongRunningExecutions, CHECK_INTERVAL, ThreadPool.Names.GENERIC);
+            }
+            return true;
+        }
+
+        public void unregister() {
+            Long previousValue = registry.remove(Thread.currentThread());
+            registered.decrementAndGet();
+            assert previousValue != null;
+            maybeLogElapsedTime(previousValue);
+        }
+
+        private void maybeLogElapsedTime(long startTime) {
+            long elapsedTime = threadPool.relativeTimeInNanos() - startTime;
+            if (elapsedTime > WARN_THRESHOLD) {
+                logger.warn(
+                    new ParameterizedMessage("Slow execution on network thread [{} milliseconds]",
+                        TimeUnit.NANOSECONDS.toMillis(elapsedTime)),
+                    new RuntimeException("Slow exception on network thread"));
+            }
+        }
+
+        private void logLongRunningExecutions() {
+            for (Map.Entry<Thread, Long> entry : registry.entrySet()) {
+                final long elapsedTime = threadPool.relativeTimeInMillis() - entry.getValue();
+                if (elapsedTime > WARN_THRESHOLD) {
+                    final Thread thread = entry.getKey();
+                    logger.warn("Slow execution on network thread [{}] [{} milliseconds]: \n{}", thread.getName(),
+                        TimeUnit.NANOSECONDS.toMillis(elapsedTime),
+                        Arrays.stream(thread.getStackTrace()).map(Object::toString).collect(Collectors.joining("\n")));
```

Comment:
I think it's because we grant

```
  // needed for top threads handling
  permission org.elasticsearch.secure_sm.ThreadPermission "modifyArbitraryThreadGroup";
```

to `codeBase "${codebase.randomizedtesting-runner}" {` right? (in `test-framework.policy`)

We actually do use the same code to get stack traces in other tests too so I'm sure it works fine with the SM (+ manually verified it).